locksmith/src/controllers/hookController.ts

```diff
@@ -0,0 +1,131 @@
+import { Response, Request } from 'express'
+import { networks } from '@unlock-protocol/networks'
+import * as z from 'zod'
+import fetch from 'cross-fetch'
+import crypto from 'crypto'
+import logger from '../logger'
+import { Hook } from '../models'
+
+const Hub = z.object({
+  topic: z.string().url(),
+  callback: z.string().url(),
+  mode: z.enum(['subscribe', 'unsubscribe']),
+  lease_seconds: z.number().optional(),
+  secret: z.string().optional(),
+})
+
+export async function subscribe(
+  hub: z.infer<typeof Hub>,
+  params: Record<string, string>
+) {
+  const hook = await Hook.findOne({
+    where: {
+      topic: hub.topic,
+      callback: hub.callback,
+    },
+  })
+
+  if (!hook) {
+    const check = isValidHubIntent(hub)
+    if (!check) {
+      const intentError = new Error(
+        `Subscriber failed to confirm intent for callback: ${hub.callback} on topic: ${hub.topic}.`
+      )
+      logger.error(intentError)
+      throw intentError
+    }
+
+    const expiration = new Date()
+    expiration.setSeconds(
+      expiration.getSeconds() + (hub.lease_seconds ?? 60 * 60 * 24 * 30)
+    )
+
+    const createdHook = await Hook.create({
+      expiration,
+      topic: hub.topic,
+      callback: hub.callback,
+      mode: hub.mode,
+      secret: hub.secret,
+      network: params.network,
+      lock: params.lock,
+    })
+
+    return createdHook
+  }
+
+  // If changing mode, reconfirm the intent and update
+  if (hook?.mode !== hub.mode) {
+    hook.mode = hub.mode
+    const check = isValidHubIntent(hub)
+    if (!check) {
+      logger.error(
+        `Subscriber failed to confirm intent for callback: ${hub.callback} on topic: ${hub.topic}.`
+      )
+      return
+    }
+    const updatedHook = await hook.save()
+    return updatedHook
+  }
+
+  return hook
+}
+
+// Check if the subscriber request is valid or not. This will post a challenge to subscriber to confirm the intent of the request.
+export async function isValidHubIntent(hub: z.infer<typeof Hub>) {
```

Comment:
This needs to happen asynchronously per https://www.w3.org/TR/websub/#x5-1-2-subscription-response-details

Reply:
I'm confused what do you mean here by happening async. I went through the spec too and I still have no idea.

Reply:
Okay, I got this should happen after responding to the subscription request with accepted or denied.

Reply:
fixed